node/src/neighborhood/mod.rs

```diff
@@ -1754,21 +2253,137 @@ mod tests {
+            .build();
+
+        let peer_actors = peer_actors_builder().build();
+        subject.handle_bind_message(BindMessage { peer_actors });
+        subject.min_hops = Hops::OneHop;
+        let exit_nodes_before_gossip = subject.init_db_countries();
+
+        subject.handle_gossip(
+            standard_gossip,
+            SocketAddr::from_str("1.1.1.1:1111").unwrap(),
+            make_cpm_recipient().0,
+        );
+
+        assert_eq!(exit_nodes_before_gossip, vec!["US".to_string()]);
+        assert_eq!(
+            subject.user_exit_preferences.db_countries,
+            vec!["FR".to_string(), "US".to_string()]
+        );
+    }
+
+    #[test]
+    fn test_with_introduction_always_have_half_neighborship_in_handle_gossip() {
+        let mut debut_subject = make_debut_subject();
+        let debut_root_key = debut_subject.neighborhood_database.root_key().clone();
+        let introducer_node = make_node_record(3333, true);
+        let introducee = make_node_record(2222, true);
+        let introducer_root_key = introducer_node.public_key().clone();
+        let mut introducer_db = debut_subject.neighborhood_database.clone();
+
+        introducer_db.set_root_key(&introducer_root_key);
+        introducer_db.add_node(introducer_node).unwrap();
+        introducer_db.add_arbitrary_half_neighbor(&introducer_root_key, &debut_root_key);
+        introducer_db.add_node(introducee.clone()).unwrap();
+        introducer_db.add_arbitrary_full_neighbor(&introducer_root_key, introducee.public_key());
+
+        let introduction_gossip = GossipBuilder::new(&introducer_db)
+            .node(&introducer_root_key, true)
+            .node(introducee.public_key(), true)
+            .build();
+        let peer_actors = peer_actors_builder().build();
+        debut_subject.min_hops = Hops::OneHop;
+        let exit_nodes_before_gossip = debut_subject.init_db_countries();
+        debut_subject.handle_bind_message(BindMessage { peer_actors });
+
+        debut_subject.handle_gossip(
+            introduction_gossip,
+            SocketAddr::from_str("3.3.3.3:3333").unwrap(),
+            make_cpm_recipient().0,
+        );
+
+        assert!(exit_nodes_before_gossip.is_empty());
+        assert_eq!(
+            debut_subject.user_exit_preferences.db_countries,
+            vec!["AU".to_string()]
+        );
+    }
+
+    #[test]
+    fn init_db_countries_works_properly() {
+        let mut subject = make_standard_subject();
+        subject.min_hops = Hops::OneHop;
+        let root_node = subject.neighborhood_database.root().clone();
+        let mut first_neighbor = make_node_record(1111, true);
+        first_neighbor.inner.country_code_opt = Some("CZ".to_string());
+        subject
+            .neighborhood_database
+            .add_node(first_neighbor.clone())
+            .unwrap();
+        subject
+            .neighborhood_database
+            .add_arbitrary_full_neighbor(root_node.public_key(), first_neighbor.public_key());
```

Comment:
Kdyz uz testujes `init_db_countries()` takhle primo (v tech predchozich testech se totiz de facto testuje taky), alespon bych otestoval ne pouze jeden extra Node, ale dal bych tam treba hned tri. Takovy test je mnohem silnejsi, nebot i dokazuje, ze tva funkce umi detekovat nekolik Nodu najednou a ne pouze jeden behem jednoho volani (Chapes, je to jen hypoteza, na ktere ti ukazuji, proc je nekdy dobre tam tech prvku pridat vic. Ja to casto delam v testech, kde jsou nejake iteratory. Nechces testovat jen s jednim, kdyz je to konstruovane na mnoho).